kotlintest-extensions/src/main/kotlin/io/kotlintest/extensions/system/SystemEnvironmentExtensions.kt

```diff
@@ -0,0 +1,191 @@
+  val originalEnvironment = System.getenv().toMap() // Using to map to guarantee it's not modified
+  
+  setEnvironmentMap(originalEnvironment overridenWith environment)
+  
+  try {
+    return block()
+  } finally {
+    setEnvironmentMap(originalEnvironment)
+  }
+}
+
+
+@PublishedApi
+internal infix fun Map<String,String>.overridenWith(map: Map<String, String?>): MutableMap<String, String> {
+  return toMutableMap().apply { putReplacingNulls(map) }
+}
+
+@PublishedApi
+// Implementation inspired from https://github.com/stefanbirkner/system-rule
+internal fun setEnvironmentMap(map: Map<String, String?>) {
+  val envMapOfVariables = getEditableMapOfVariables()
+  val caseInsensitiveEnvironment = getCaseInsensitiveEnvironment()
+  
+  envMapOfVariables.clear()
+  caseInsensitiveEnvironment?.clear()
+  
+  envMapOfVariables.putReplacingNulls(map)
+  caseInsensitiveEnvironment?.putReplacingNulls(map)
+}
+
+@Suppress("UNCHECKED_CAST")
+private fun getEditableMapOfVariables(): MutableMap<String, String> {
+  val systemEnv = System.getenv()
+  val classOfMap = systemEnv::class.java
+  
+  return classOfMap.getDeclaredField("m").asAccessible().get(systemEnv) as MutableMap<String, String>
+}
+
+@Suppress("UNCHECKED_CAST")
+private fun getCaseInsensitiveEnvironment(): MutableMap<String, String>? {
+  val processEnvironmentClass = Class.forName("java.lang.ProcessEnvironment")
+  
+  return try {
+    processEnvironmentClass.getDeclaredField("theCaseInsensitiveEnvironment").asAccessible().get(null) as MutableMap<String, String>?
+  } catch (e: NoSuchFieldException) {
+    // Only available in Windows, ok to return null if it's not found
+    null
+  }
+}
+
+private fun  Field.asAccessible(): Field {
+  return apply { isAccessible = true }
+}
+
+private fun MutableMap<String,String>.putReplacingNulls(map: Map<String, String?>) {
+  map.forEach { key, value ->
+    if(value == null) remove(key) else put(key, value)
+  }
+}
+
+
+/**
+ * Overrides System Environment with chosen keys and values
+ *
+ * This is a Listener for code that uses Environment Variables. It overrides the specific keys from [System.getenv]
+ * with the specified values, only during the execution of a test.
+ *
+ * To do this, this listener uses a trick that makes the System Environment editable, and overrides the keys. Any previous
+ * environment (anything not overridden) will also be in the environment. If the chosen key is in the environment,
+ * it will be overridden. If the chosen key is not in the environment, it will be included.
+ *
+ * After the execution of the test, the environment is set to what it was before.
+ *
+ * **ATTENTION**: This code is susceptible to race conditions. If you attempt to change the environment while it was
+ * already changed, the result is inconsistent, as the System Environment Map is a single map.
+ */
+class SystemEnvironmentListener(private val environment: Map<String, String>) : TestListener {
```

Comment:
This is all great except I would make two versions.

SystemEnvironmentTestListener

SystemEnvironmentProjectListener.

The second one only sets the env at the start of the project, and resets at the end. This means you can use parallelism with it.
The first one is when you only want to override for one or two tests or something.

Reply:
I didn't want to provide two versions as to not confuse users. But this makes sense. I"ll change it

Reply:
How to test for project configuration tho?

Reply:
You don't need to test for anything, just need to have two versions which use different callback methods. The user can decide which to use.